src/bp/ui-admin/src/telemetry/client.ts

```diff
@@ -0,0 +1,143 @@
+import ms from 'ms'
+import uuid from 'uuid'
+
+import store from '../store'
+
+interface EventPackageInfoType {
+  [key: string]: {
+    locked: boolean
+    timeout: string
+    getPackage: Function
+  }
+}
+
+export const axiosConfig = {
+  baseURL: window.TELEMETRY_URL,
+  headers: {
+    withCredentials: false
+  }
+}
+
+const dataSchemaVersion = '1.0.0'
+
+const telemetrySchemaVersion = '1.0.0'
+
+const storeInfos: Function[] = []
+
+const eventPackageInfo: EventPackageInfoType = {}
+
+const toggleLock = (eventName: string) => {
+  if (_.has(eventPackageInfo, eventName)) {
+    eventPackageInfo[eventName].locked = !eventPackageInfo[eventName].locked
+  }
+}
+
+const getEventLockTimeout = (event: string) => {
+  const item = window.localStorage.getItem(event)
+  if (item !== null) {
+    const timeout = parseInt(item) - new Date().getTime()
+    if (timeout > 0) {
+      return timeout
+    }
+    window.localStorage.removeItem(event)
+  }
+  return 0
+}
+
+const addEventLockTimeout = (event: string, timeout: number) => {
+  setTimeout(() => toggleLock(event), timeout)
+  window.localStorage.setItem(event, (timeout + moment().valueOf()).toString())
+}
+
+const checkStoreInfoReceived = () => !storeInfos.find(info => info() === undefined)
+
+export const trackReduxStoreInfo = (pathInStore: string) => {
+  storeInfos.push(() => {
+    return _.get(store.getState(), pathInStore)
+  })
+}
+
+export const addTelemetryEvent = (name: string, timeout: string, getPackage: Function) => {
+  eventPackageInfo[name] = {
+    locked: getEventLockTimeout(name) >= 0,
+    timeout,
+    getPackage
+  }
+}
+
+const checkTelemetry = async () => {
+  for (const eventName in eventPackageInfo) {
+    if (!eventPackageInfo[eventName].locked) {
+      toggleLock(eventName)
+
+      try {
+        const pkg = eventPackageInfo[eventName].getPackage()
+
+        if (typeof pkg === 'object') {
+          await sendTelemetryEvent(pkg, eventName)
+        } else {
+          console.error('The package received was incorrect', pkg)
+        }
+      } catch (err) {
+        console.error(`Could not send the telemetry package to the storage server`, err)
+      }
+    }
+  }
+}
+
+export const startTelemetry = () => {
+  trackReduxStoreInfo('user.profile.email')
+
+  trackReduxStoreInfo('version.currentVersion')
+
+  trackReduxStoreInfo('license.licensing.isPro')
+
+  addTelemetryEvent('ui_language', '8h', () => {
+    return {
+      user: {
+        email: store.getState().user.profile.email,
+        timezone: Intl.DateTimeFormat().resolvedOptions().timeZone
+      },
+      language: lang.getLocale()
+    }
+  })
+
+  for (const event in eventPackageInfo) {
+    if (getEventLockTimeout(event) >= 0) {
+      setTimeout(() => toggleLock(event), getEventLockTimeout(event))
+    }
+  }
+
+  setInterval(() => {
+    if (checkStoreInfoReceived() && window.TELEMETRY_URL) {
+      checkTelemetry().catch(err => {
+        console.error(err)
+      })
+    }
+  }, ms('30s'))
+}
```

Comment:
Drop all that locking logic. Unless you have a very good explanation why it's absolutely necessary. I'm absolutely sure you can find a better and simpler way to do this

Reply:
I made the locking mechanism to avoid sending too many packages to the server. The "ui_language" is not an info that changes often on the frontend. So this is why this package is sent only every 8 hours at best.

Reply:
Just put an expiry date in the local storage (one JSON object for any new packages, no need to bloat it). Then when the browser is refreshed, check if the expiry is past, then send the package.

There's absolutely no gain in having an interval running permanently to send the next package at precisely 8 hours later. Just run the check on first pageload (setTimeout), then act accordingly. No locking mechanism needed

Reply:
Fixed in :
52d479c?